### DISCUSSION TITLE
NextAuth example works locally but fails on Zeit

### DISCUSSION THREAD

**Discussion by shea256** [vercel/next.js#11530] [Help] [answered]
Hey all,

In the NextAuth example (https://github.com/iaincollins/next-auth/tree/master/example), the file `pages/auth/index.js` has the following code:

```js
import React from 'react'
import Router from 'next/router'
import { NextAuth } from 'next-auth/client'
import Layout from '../../components/layout'

export default class extends React.Component {
  
  static async getInitialProps({req}) {
    return {
      session: await NextAuth.init({req}),
      linkedAccounts: await NextAuth.linked({req}),
      providers: await NextAuth.providers({req})
    }
  }

  render() {
    ...
  }
}
```

This page works perfectly fine when running locally, but unfortunately fails when deployed to Zeit using now.

In the deployment, I get the following error:

<img width="1161" alt="Screen Shot 2020-03-31 at 12 31 54 PM" src="https://user-images.githubusercontent.com/994646/78070747-53ed2700-736a-11ea-9efb-9819866578c4.png">

I tried troubleshooting this to no avail.

I saw this note in the NextAuth documentation:

> CSRF, Session and Linked Account endpoints are provided for Single Page Apps.

Based on this, I had a suspicion that `NextAuth.linked` and `NextAuth.providers` shouldn't be called on the server and should only be called on the client, so I modified the code to only call these functions on the client.

Unfortunately, this didn't fix the problem. The page still shows a server error when it is loaded on the client.

Would appreciate any help, and would be great to find out if this is an issue with NextAuth, Next.js, Zeit, or one of my minor modifications.

Thanks!

**Comment by lfades** [comment]
This is an issue with `next-auth`. That repo is not maintained or recommended by us so please create an issue in their repo instead :pray:

**Reply by shea256** [reply]
OK thanks @lfades will do.

And if you don't recommend `next-auth` then what do you recommend for OAuth authentication with next.js?

**Reply by shea256** [reply]
Also @lfades if the app works perfectly fine locally but fails on Zeit, it may be the case that there is something wrong with Zeit serverless functions.

What makes you so certain it is `next-auth`?

**Reply by shea256** [reply]
@lfades FYI @rauchg is the one who told me to post here to start a discussion.

**Reply by shea256** [reply]
For reference, here is the cross-posting on the `next-auth` repo: https://github.com/iaincollins/next-auth/issues/96

**Reply by lfades** [reply]
We don't recommend `next-auth` for multiple reasons, one of those is that it's not compatible with ZEIT Now, it's outdated and it currently uses a custom server, which makes it incompatible.

We're also very into security, and because of that the example we currently recommend is the one with [Auth0](https://github.com/zeit/next.js/tree/canary/examples/auth0). Their company is very good at doing proper and secure authentication. We can't guarantee that with external packages.

**Reply by lfades** [reply]
In summary, `next-auth` can be updated to support serverless functions instead of a custom server, but that's currently not the case. To handle authentication check the [Auth0 example](https://github.com/zeit/next.js/tree/canary/examples/auth0) I shared before, in where we worked together with Auth0.

We also have a [passport example](https://github.com/zeit/next.js/tree/canary/examples/with-passport) which shows very basic cookie authentication with API Routes in a statically optimized application, but it can't compare to the security added by Auth0.

**Reply by shea256** [reply]
@lfades Ah so it seems the custom server they have is the culprit. Thank you, that's very helpful. I will give the auth0 example a spin.

**Reply by shea256** [reply]
So I tried out this example and it seems ok. Unfortunately it requires a redirect to the Auth0 domain *and then* the domain of the OAuth provider and it requires you to use the full Auth0 auth window.

**Comment by iaincollins** [comment, marked as answer]
Hi @shea256

I am the lead maintainer of [next-auth](https://next-auth.js.org)

I can confirm, as @lfades says, that this problem is indeed an issue `next-auth` v1 and not with to Next.js or now.sh. NextAuth.js v1 was created for now.sh v1 and is not compatible with now.sh from Vercel today and does not take advantage of features in recent versions of Next.js.

There is some good news though!

You might want to check out [next-auth.js.org](https://next-auth.js.org) for the beta of v2 or check out [next-auth-example.now.sh](https://next-auth-example.now.sh) for a demo.

* Supports Next.js 9+ and Serverless
* _Much easier_ to set up and use than v1
* Supports [many popular providers](https://next-auth.js.org/options/providers), including [Auth0](https://next-auth.js.org/providers/auth0)
* Install it will `npm next-auth@beta` 

NextAuth.js v2 is an open source community project. It's currently in beta and the current release candidate is looking good so far and we expect to release v2 out of beta in later this week. 🤞
